File: src/jv/ItemContainer.java
package jv;

import java.util.Arrays;
import jv.item.BaseItem;
import jv.item.INVALID_ITEM;

import static jv.ArrayFunctional.*;

public class ItemContainer {
    public final BaseItem[] wrapped;
    public ItemContainer(BaseItem[] wrapped) {
        this.wrapped = new BaseItem[wrapped.length];
        System.arraycopy(wrapped, 0, this.wrapped, 0, wrapped.length);
    }

    public final void fastShift() {
        if (wrapped.length == 0 || wrapped.length == 1) return;
        forEachIndexed(
                partitionJoin(this.wrapped, baseItem -> !baseItem.isInvalidItem()),
                (index, item) -> this.wrapped[index] = item);
    }

    public final void memoryFriendlyShift() {
        if (wrapped.length == 0 || wrapped.length == 1) return;
        int nextSlotToFill = -1;
        int currentSlot = 0;
        BaseItem temporaryItem;
        while (currentSlot < wrapped.length) {
            temporaryItem = wrapped[currentSlot];
            if (temporaryItem.isInvalidItem()) {
                if (nextSlotToFill < 0) {
                    nextSlotToFill = currentSlot;
                }
                currentSlot++;
                continue;
            }
            if (nextSlotToFill >= 0) {
                wrapped[nextSlotToFill] = wrapped[currentSlot];
                wrapped[currentSlot] = INVALID_ITEM.getInstance();
                currentSlot = nextSlotToFill + 1;
                nextSlotToFill = -1;
                continue;
            }
            currentSlot ++;
        }
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) return true;
        if (!(obj instanceof ItemContainer)) return false;
        return Arrays.equals(wrapped, ((ItemContainer) obj).wrapped);
    }

    @Override
    public String toString() { return Arrays.toString(wrapped); }
}

File: src/kt/ItemContainer.kt
package kt

class ItemContainer(wrapped: Array<BaseItem>) {

    val wrapped = wrapped.copyOf()

    /**
     * This is essentially equal speed until around 500 items in container
     */
    fun memoryFriendlyShift() {
        if (wrapped.isEmpty() || wrapped.size == 1) return
        var nextSlotToFill = -1
        var currentSlot = 0
        var temporaryItem: BaseItem
        while (currentSlot < wrapped.size) {
            temporaryItem = wrapped[currentSlot]
            if (temporaryItem.isInvalidItem) {
                if (nextSlotToFill < 0) {
                    nextSlotToFill = currentSlot
                }
                currentSlot++
                continue
            }
            if (nextSlotToFill >= 0) {
                wrapped[nextSlotToFill] = wrapped[currentSlot]
                wrapped[currentSlot] = INVALID_ITEM
                currentSlot = nextSlotToFill + 1
                nextSlotToFill = -1
                continue
            }
            currentSlot ++
        }
    }

    fun fastShift() {
        if (wrapped.isEmpty() || wrapped.size == 1) return
        wrapped
            .partitionJoin { !it.isInvalidItem }
            .forEachIndexed { slotIndex, item -> wrapped[slotIndex] = item }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ItemContainer

        if (!wrapped.contentEquals(other.wrapped)) return false

        return true
    }

    override fun hashCode(): Int = wrapped.contentHashCode()

    override fun toString() = wrapped.contentToString()
}